instrumentation/opentelemetry-instrumentation-httpx/src/opentelemetry/instrumentation/httpx/__init__.py

```diff
@@ -788,67 +959,90 @@ def instrument_client(
             response_hook: A hook that receives the span, request, and response
                 that is called right before the span ends
         """
-        # pylint: disable=protected-access
-        if not hasattr(client, "_is_instrumented_by_opentelemetry"):
-            client._is_instrumented_by_opentelemetry = False
 
-        if not client._is_instrumented_by_opentelemetry:
-            if isinstance(client, httpx.Client):
-                client._original_transport = client._transport
-                client._original_mounts = client._mounts.copy()
-                transport = client._transport or httpx.HTTPTransport()
-                client._transport = SyncOpenTelemetryTransport(
-                    transport,
-                    tracer_provider=tracer_provider,
-                    request_hook=request_hook,
-                    response_hook=response_hook,
-                )
-                client._is_instrumented_by_opentelemetry = True
-                client._mounts.update(
-                    {
-                        url_pattern: (
-                            SyncOpenTelemetryTransport(
-                                transport,
-                                tracer_provider=tracer_provider,
-                                request_hook=request_hook,
-                                response_hook=response_hook,
-                            )
-                            if transport is not None
-                            else transport
-                        )
-                        for url_pattern, transport in client._original_mounts.items()
-                    }
-                )
+        if getattr(client, "_is_instrumented_by_opentelemetry", False):
+            _logger.warning(
+                "Attempting to instrument Httpx client while already instrumented"
+            )
+            return
 
-            if isinstance(client, httpx.AsyncClient):
-                transport = client._transport or httpx.AsyncHTTPTransport()
-                client._original_mounts = client._mounts.copy()
-                client._transport = AsyncOpenTelemetryTransport(
-                    transport,
-                    tracer_provider=tracer_provider,
+        _OpenTelemetrySemanticConventionStability._initialize()
+        sem_conv_opt_in_mode = _OpenTelemetrySemanticConventionStability._get_opentelemetry_stability_opt_in_mode(
+            _OpenTelemetryStabilitySignalType.HTTP,
+        )
+        tracer = get_tracer(
+            __name__,
+            instrumenting_library_version=__version__,
+            tracer_provider=tracer_provider,
+            schema_url=_get_schema_url(sem_conv_opt_in_mode),
+        )
+
+        if iscoroutinefunction(request_hook):
```

Comment:
Not sure if we want to introduce this behavior (auto setting request hook to async if it is indeed an async function). What happens if they are both are provided as an async function? Wouldn't it be confusing if `request_hook` overrides `async_request_hook`?

Reply:
This was already the behavior of `instrument_client`, agree that it's not great that it behaves differently than plain `instrument`. See old code at https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2909/files/f0880e65765f01f761bd3c2ff691e6414c53d4f2#diff-eed55d62a8d5f01b055c1bb2d63934c48cfd767416c03826f2c19d4442a76e7bL839